### PR TITLE
aabtree and aabtreebuilder

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/w3d/renderer/aabtreebuilder.cpp [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+    newnode->Max.x = node->m_max.X;
+    newnode->Max.y = node->m_max.Y;
+    newnode->Max.z = node->m_max.Z;
+
+    if (node->m_front != nullptr) {
+        captainslog_assert(node->m_back != nullptr);
+        newnode->FrontOrPoly0 = node->m_front->m_index;
+        newnode->BackOrPolyCount = node->m_back->m_index;
+    } else {
+        newnode->FrontOrPoly0 = cur_poly | 0x80000000;
```
Comment (tomsons26) [inline src/w3d/renderer/aabtreebuilder.cpp]: Is there some context we can add for 0x80000000? Don't like magic numbers

File: src/w3d/renderer/aabtreebuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+void AABTreeBuilderClass::Compute_Bounding_Box(CullNodeStruct *node)
+{
+    if (node->m_front) {
+        Compute_Bounding_Box(node->m_front);
+    }
+
+    if (node->m_back) {
+        Compute_Bounding_Box(node->m_back);
+    }
+
```
Comment (tomsons26) [inline src/w3d/renderer/aabtreebuilder.cpp:313]: NO_MAX_SCREEN_SIZE def isn't used here, bfme2 shows it was called really_big, possibly constant in function?

File: src/w3d/renderer/aabtreebuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+        if (node->m_back->m_min.Z < node->m_min.Z) {
+            node->m_min.Z = node->m_back->m_min.Z;
+        }
+
+        if (node->m_back->m_max.Z > node->m_max.Z) {
+            node->m_max.Z = node->m_back->m_max.Z;
+        }
+    }
+
+    captainslog_assert(node->m_min.X != 100000.0f);
```
Comment (tomsons26) [inline src/w3d/renderer/aabtreebuilder.cpp]: this checked said really_big

File: src/w3d/renderer/aabtreebuilder.cpp [hunk 4]
```diff
@@ -0,0 +1,569 @@
...
+
+        mask |= ON;
+    }
+
+    if (mask == ON) {
+        return ON;
+    }
+
+    if ((mask & ~(POS | ON)) == 0) {
+        return POS;
```
Comment (tomsons26) [inline src/w3d/renderer/aabtreebuilder.cpp]: FRONT

File: src/w3d/renderer/aabtreebuilder.cpp [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+    }
+
+    if ((mask & ~(NEG | ON)) == 0) {
+        return NEG;
```
Comment (tomsons26) [inline src/w3d/renderer/aabtreebuilder.cpp]: BACK

File: src/w3d/renderer/aabtree.cpp [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+    if (boxtest.Cull(node->m_min, node->m_max)) {
+        return false;
+    }
+
+    bool res = false;
+
+    if (node->Is_Leaf()) {
+        return Cast_OBBox_To_Polys(node, boxtest);
+    } else {
+        res = res | Cast_OBBox_Recursive(&(m_nodes[node->Get_Front_Child()]), boxtest);
```
Comment (tomsons26) [inline src/w3d/renderer/aabtree.cpp]: |=

File: src/w3d/renderer/aabtree.cpp [hunk 7]
```diff
@@ -0,0 +1,668 @@
...
+    if (boxtest.Cull(node->m_min, node->m_max)) {
+        return false;
+    }
+
+    bool res = false;
+
+    if (node->Is_Leaf()) {
+        return Cast_AABox_To_Polys(node, boxtest);
+    } else {
+        res = res | Cast_AABox_Recursive(&(m_nodes[node->Get_Front_Child()]), boxtest);
```
Comment (tomsons26) [inline src/w3d/renderer/aabtree.cpp]: |=

File: src/w3d/renderer/aabtree.cpp [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+bool AABTreeClass::Cast_AABox_To_Polys(CullNodeStruct *node, AABoxCollisionTestClass &boxtest)
+{
+    if (node->Get_Poly_Count() > 0) {
+        TriClass tri;
+
+        const Vector3 *loc = m_mesh->Get_Vertex_Array();
+        const TriIndex *polyverts = m_mesh->Get_Polygon_Array();
+
+        int polyhit = -1;
+        int poly0 = node->Get_Poly0();
```
Comment (tomsons26) [inline src/w3d/renderer/aabtree.cpp:402]: move to top to keep consistent to below

File: src/w3d/renderer/aabtree.cpp [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+                v0, v1, v2, tri_plane, start_point, axis_r, axis_1, axis_2, direction, flags);
+        }
+    }
+
+    return count;
+}
+
+bool AABTreeClass::Cast_AABox_To_Polys(CullNodeStruct *node, AABoxCollisionTestClass &boxtest)
+{
+    if (node->Get_Poly_Count() > 0) {
```
Comment (tomsons26) [inline src/w3d/renderer/aabtree.cpp]: check polycount

File: src/w3d/renderer/aabtree.cpp [hunk 10]
```diff
@@ -0,0 +1,668 @@
...
+    if (raytest.Cull(node->m_min, node->m_max)) {
+        return false;
+    }
+
+    bool res = false;
+
+    if (node->Is_Leaf()) {
+        return Cast_Ray_To_Polys(node, raytest);
+    } else {
+        res = res | Cast_Ray_Recursive(&(m_nodes[node->Get_Front_Child()]), raytest);
```
Comment (tomsons26) [inline src/w3d/renderer/aabtree.cpp]: |=

File: src/w3d/renderer/aabtree.cpp [hunk 11]
```diff
@@ -0,0 +1,668 @@
...
+    if (test.Cull(node->m_min, node->m_max)) {
+        return false;
+    }
+
+    bool res = false;
+
+    if (node->Is_Leaf()) {
+        return Intersect_OBBox_With_Polys(node, test);
+    } else {
+        res = res | Intersect_OBBox_Recursive(&(m_nodes[node->Get_Front_Child()]), test);
```
Comment (tomsons26) [inline src/w3d/renderer/aabtree.cpp]: |=

File: src/w3d/renderer/meshgeometry.cpp [hunk 12]
```diff
@@ -27,11 +34,128 @@ const Vector3 *MeshGeometryClass::Get_Vertex_Normal_Array()
...
+    m_boundSphereRadius *= max;
+
+    if (sc.X != sc.Y || sc.Y != sc.Z) {
+        Set_Flag(DIRTY_VNORMALS, true);
+    }
+
+    Set_Flag(DIRTY_PLANES, true);
+
+    if (m_cullTree) {
+        if (GameMath::Fabs(sc.X - sc.Y) >= 0.00009999999747378752f
```
Comment (tomsons26) [inline src/w3d/renderer/meshgeometry.cpp]: flip the if, and the double is MATH_EPSILON

File: src/w3d/renderer/aabtree.cpp (hunk 10 above)
Comment (jonwil) [reply src/w3d/renderer/aabtree.cpp]: Fixed.

File: src/w3d/renderer/aabtree.cpp (hunk 7 above)
Comment (jonwil) [reply src/w3d/renderer/aabtree.cpp]: Fixed.

File: src/w3d/renderer/aabtree.cpp (hunk 6 above)
Comment (jonwil) [reply src/w3d/renderer/aabtree.cpp]: Fixed.

File: src/w3d/renderer/aabtree.cpp (hunk 11 above)
Comment (jonwil) [reply src/w3d/renderer/aabtree.cpp]: Fixed.

File: src/w3d/renderer/aabtree.cpp (hunk 8 above)
Comment (jonwil) [reply src/w3d/renderer/aabtree.cpp:402]: Fixed.

File: src/w3d/renderer/meshgeometry.cpp (hunk 12 above)
Comment (jonwil) [reply src/w3d/renderer/meshgeometry.cpp]: Fixed.

File: src/w3d/renderer/aabtreebuilder.cpp (hunk 1 above)
Comment (jonwil) [reply src/w3d/renderer/aabtreebuilder.cpp]: Fixed.

File: src/w3d/renderer/aabtreebuilder.cpp (hunk 2 above)
Comment (jonwil) [reply src/w3d/renderer/aabtreebuilder.cpp:313]: Fixed.

File: src/w3d/renderer/aabtreebuilder.cpp (hunk 3 above)
Comment (jonwil) [reply src/w3d/renderer/aabtreebuilder.cpp]: Fixed.

File: src/w3d/renderer/aabtree.cpp (hunk 9 above)
Comment (jonwil) [reply src/w3d/renderer/aabtree.cpp]: Fixed.

File: src/w3d/renderer/aabtreebuilder.cpp (hunk 5 above)
Comment (jonwil) [reply src/w3d/renderer/aabtreebuilder.cpp]: Fixed.

File: src/w3d/renderer/aabtreebuilder.cpp (hunk 4 above)
Comment (jonwil) [reply src/w3d/renderer/aabtreebuilder.cpp]: Fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**